Basic function support

DiegoWearden opened the pull request:
## Added function support

### Implementation details:
- Term.hs: added `Fun String Term` and `App Term Term`
- Value.hs: added `ClosureVal String Term [(String, Value)]` (left captures unused for now)
- Small.hs:
  - `Fun` constructs a function value, reduces to `ClosureVal x body []`
  - `App` applies a function with call by value semantics (evaluate function, then argument, then apply)
  - Prevent environment leakage by evaluating the body on a saved/restored machine state with `x = arg`

### Here's an example:

```haskell
let inc = Fun "x" (BinaryOps Add (Var "x") (Literal 1))
App inc (Literal 41)  -- outputs 42
```

### Some limitations right now:
- I only implemented single parameter functions (although you can technically curry/nest for multiple parameters)
- functions don’t capture outer variables, I left scoping for later

Emerald-Z commented on Term.hs:
```diff
@@ -20,4 +20,6 @@ data Term
   | Write Term
   | BoolLit Bool
+  | Fun String Term
+  | App Term Term
```
Lowkey a nit pick so feel free to let me know, but maybe we should call this something more readable for later, like ApplyFun just for clarity!

Emerald-Z reviewed (commented):
Looks good! Left one small comment, and I also agree with the multi-args comment made earlier. I think it's cool if you don't want to include it in this particular PR, just make that clear!

Emerald-Z commented:
Also might be a dumb question, but are we supporting functions with no arguments?

CinnamonNinian reviewed (approved):
It looks like you've implemented multiple argument functions without the need for currying. Nicely done!

DiegoWearden commented:
@Emerald-Z @amerikrainian Thanks for your comments! I have updated the code to support functions with no arguments and multiple arguments:

## Extended function support

### Implementation details:
- Term.hs: added `Fun [String] Term` and `ApplyFun Term [Term]`
- Value.hs: added `ClosureVal [String] Term [(String, Value)]` (captures store applied args, still no scoping yet)
- Small.hs:
  - `Fun` constructs a function value, reduces to `ClosureVal xs body []`
  - `ApplyFun` uses call-by-value, left-to-right (evaluate function, then each argument, then apply)
  - While stepping we keep the original call shape. When it’s time to run, we temporarily set each parameter name to its argument value, run the body once, then restore things to how they were.
  - Multi-parameter: we evaluate the function, then each argument left to right. Each argument fills the next parameter. If you provide fewer arguments, you get back a function that remembers what you gave (partial application, example of how this works is below). Once you’ve provided them all, we run the body once with those values.
  - Zero-parameter: there are no parameters to fill, so calling it just runs the body. Passing any arguments is an error. Calling a function that still expects parameters without supplying them is also an error.

### Here's an example:

```haskell
-- single parameter
let inc = Fun ["x"] (BinaryOps Add (Var "x") (Literal 1))
ApplyFun inc [Literal 41]  -- outputs 42
```

```haskell
-- zero parameters
let fortyTwo = Fun [] (Literal 42)
ApplyFun fortyTwo []  -- outputs 42
```

```haskell
-- partial application
let add = Fun ["x","y"] (BinaryOps Add (Var "x") (Var "y"))
let add2 = ApplyFun add [Literal 2]       -- returns a function waiting for y
ApplyFun add2 [Literal 40]                -- outputs 42
```

### Some notes and limitations:
- Functions do not capture outer variables (still no scoping yet)

amerikrainian reviewed (approved):
Looks good to me for without scopes and should be mostly fine for when those get added; we'll just change what happens when entering/exiting functions.
